File: Apps/iOS/Domain/Entities/Scores/ScoreWriting.swift
extension ScoreEvent {
    static func blow(_ hole: Hole, _ beats: Double, vibrato: Double = 0, slideFrom: Hole? = nil) -> ScoreEvent {
        blow([hole], beats, vibrato: vibrato, slideFrom: slideFrom)
    }

    static func blow(
        _ holes: [Hole],
        _ beats: Double,
        vibrato: Double = 0,
        slideFrom: Hole? = nil
    ) -> ScoreEvent {
        .note(
            ScoreNote(
                holes: holes,
                breath: .blow,
                beats: beats,
                vibrato: vibrato,
                slideFrom: slideFrom
            )
        )
    }

    static func draw(
        _ hole: Hole,
        _ beats: Double,
        bentBy semitones: Double = 0,
        releasingTo released: Double? = nil,
        vibrato: Double = 0,
        slideFrom: Hole? = nil,
        shakenWith shaken: Hole? = nil
    ) -> ScoreEvent {
        .note(
            ScoreNote(
                holes: [hole],
                breath: .draw,
                beats: beats,
                bentBySemitones: semitones,
                vibrato: vibrato,
                slideFrom: slideFrom,
                shakenWith: shaken,
                bendEndsAtSemitones: released
            )
        )
    }

    static func draw(
        _ holes: [Hole],
        _ beats: Double,
        vibrato: Double = 0,
        slideFrom: Hole? = nil
    ) -> ScoreEvent {
        .note(
            ScoreNote(
                holes: holes,
                breath: .draw,
                beats: beats,
                vibrato: vibrato,
                slideFrom: slideFrom
            )
        )
    }

    static func overblow(_ hole: Hole, _ beats: Double) -> ScoreEvent {
        .note(ScoreNote(holes: [hole], breath: .blow, beats: beats, isOverbent: true))
    }
}

File: Web/src/domain/scores/scoreWriting.ts
import type { Hole } from '../instrument/hole.js'
import type { ScoreEvent, ScoreNote } from './score.js'

export interface Expression {
    readonly bentBy?: number
    readonly releasingTo?: number
    readonly vibrato?: number
    readonly slideFrom?: Hole
    readonly shakenWith?: Hole
}

export function blow(holes: Hole | readonly Hole[], beats: number, expression: Expression = {}): ScoreEvent {
    return played(holes, 'blow', beats, expression)
}

export function draw(holes: Hole | readonly Hole[], beats: number, expression: Expression = {}): ScoreEvent {
    return played(holes, 'draw', beats, expression)
}

export function overblow(hole: Hole, beats: number): ScoreEvent {
    return { kind: 'note', note: { holes: [hole], breath: 'blow', beats, isOverbent: true } }
}

export function rest(beats: number): ScoreEvent {
    return { kind: 'rest', beats }
}

function played(
    holes: Hole | readonly Hole[],
    breath: 'blow' | 'draw',
    beats: number,
    expression: Expression
): ScoreEvent {
    const note: ScoreNote = {
        holes: Array.isArray(holes) ? holes : [holes as Hole],
        breath,
        beats,
        bentBySemitones: expression.bentBy ?? 0,
        vibrato: expression.vibrato ?? 0,
        ...(expression.slideFrom === undefined ? {} : { slideFrom: expression.slideFrom }),
        ...(expression.shakenWith === undefined ? {} : { shakenWith: expression.shakenWith }),
        ...(expression.releasingTo === undefined ? {} : { bendEndsAtSemitones: expression.releasingTo })
    }
    return { kind: 'note', note }
}
